hail/python/hail/expr/functions.py

```diff
@@ -2487,7 +2487,13 @@ def downcode(c, i) -> CallExpression:
     Preserve the third allele and downcode all other alleles to reference.
 
     >>> hl.eval(hl.downcode(hl.call(1, 2), 2))
-    Call(alleles=[0, 2], phased=False)
```

Comment:
Why did this not fail the tests?

Reply:
we don't test against output in Python docs. Only a small fraction of our codebase could possibly be tested that way, and it would take some work to do that.

Reply:
Sorry, I'm gonna push back on this. I modified `hail/python/hail/docs/doctest.sh` as follows:

```
diff --git a/hail/python/hail/docs/doctest.sh b/hail/python/hail/docs/doctest.sh
index 4e0e417fb..9572ec645 100755
--- a/hail/python/hail/docs/doctest.sh
+++ b/hail/python/hail/docs/doctest.sh
@@ -2,6 +2,8 @@
 
 set -e
 
+(cd python/hail/ && python -m doctest expr/functions.py)
+
 # clean directories
 rm -rf build/tmp/doctest
```

and I see 52 errors when I run `./gradlew doctest`

```
... detailed error info above ...
**********************************************************************
41 items had failures:
   1 of   4 in functions.any
   1 of   3 in functions.argmax
   1 of   3 in functions.argmin
   1 of   2 in functions.array
   1 of   2 in functions.bind
   1 of   1 in functions.call
   1 of   1 in functions.dict
   1 of   1 in functions.downcode
   1 of   2 in functions.entropy
   1 of   4 in functions.filter
   2 of   4 in functions.find
   2 of   2 in functions.fisher_exact_test
   1 of   3 in functions.float32
   2 of   2 in functions.interval
   1 of   1 in functions.is_complex
   1 of   4 in functions.is_finite
   1 of   4 in functions.is_infinite
   1 of   3 in functions.is_nan
   1 of   1 in functions.is_star
   1 of   2 in functions.json
   1 of   3 in functions.literal
   1 of   3 in functions.locus_from_global_position
   1 of   1 in functions.locus_interval
   1 of   2 in functions.mean
   1 of   7 in functions.mendel_error_code
   2 of   2 in functions.min_rep
   2 of   2 in functions.null
   1 of   2 in functions.or_missing
   1 of   1 in functions.parse_call
   2 of   2 in functions.parse_locus_interval
   1 of   1 in functions.parse_variant
   2 of   2 in functions.rand_beta
   1 of   2 in functions.rand_bool
   2 of   2 in functions.rand_dirichlet
   2 of   2 in functions.rand_gamma
   2 of   2 in functions.rand_norm
   1 of   2 in functions.rand_pois
   2 of   2 in functions.rand_unif
   1 of   2 in functions.set
   1 of   4 in functions.sorted
   1 of   1 in functions.str
***Test Failed*** 52 failures.
```

Some of these are silly doc test things (we have to say `expr is None` since `None` is not printed by the interpreter) but there's some serious bugs in here!

The downcode you're fixing here:
```
File "expr/functions.py", line 36, in functions.downcode
Failed example:
    hl.eval(hl.downcode(hl.call(1, 2), 2))
Expected:
    Call(alleles=[0, 2], phased=False)
Got:
    Call(alleles=[0, 1], phased=False)
```
negative array indexing is broken
```
File "expr/functions.py", line 41, in functions.filter
Failed example:
    hl.eval(hl.filter(lambda x: ~(x[-1] == 'e'), s))
Expected:
    {'Bob'}
Got:
    {'Alice', 'Charlie', 'Bob'}
```
```
File "expr/functions.py", line 38, in functions.find
Failed example:
    hl.eval(hl.find(lambda x: x[-1] == 'x', a))
Expected:
    'fox'
Got nothing
```

`hl.is_deletion` is not as documented
```
File "expr/functions.py", line 35, in functions.is_complex
Failed example:
    hl.eval(hl.is_deletion('ATT', 'GCA'))
Expected:
    True
Got:
    False
```
```
File "expr/functions.py", line 35, in functions.is_star
Failed example:
    hl.eval(hl.is_deletion('A', '*'))
Expected:
    True
Got:
    False
```

Apparently rand_beta can return `nan`?
```
File "expr/functions.py", line 45, in functions.rand_beta
Failed example:
    hl.eval(hl.rand_beta(0, 1))
Expected:
    0.6696807666871818
Got:
    nan
```

(there's a few randomness errors, but now that randomness is seeded we should be able to fix values right?)

The docs are wrong for what `reverse=False` does
```
File "expr/functions.py", line 40, in functions.sorted
Failed example:
    hl.eval(hl.sorted(a, reverse=False))
Expected:
    ['Charlie', 'Bob', 'Alice']
Got:
    ['Alice', 'Bob', 'Charlie']
```

We should clean up these tests and use doctest!

Reply:
wow, nice work. This is a mess.

Reply:
I'll open an issue and assign it to myself. Will work on that right now.